New parallel python tutotials (#744)
NREL/HPC: docs/Documentation/Development/Languages/Python/KestrelParallelPythonJupyter/pyEnvsAndLaunchingJobs.md

---
layout: default

> title: Kestrel Parallel Python Jupyter

Reviewer: ```suggestion
title: Parallel Python with Jupyter
```

parent: Python
nav_order: 2
has_children: true
grand_parent: Programming Languages
---

> # Interactive Parallel Python on Kestrel with Jupyter

Reviewer: ```suggestion
# Interactive Parallel Python with Jupyter
```

> For a general introduction to using Jupyter notebooks on Kestrel, please refer to the [official documentation](https://nrel.github.io/HPC/Documentation/Development/Jupyter/). This page covers how to leverage parallel computing with python through Jupyter notebooks on compute nodes. Accompanying notebooks to test if the environment and the job have been configured correctly are under the topic 'Parallel Interactive Tutorials' on the main [Python](../index.md/#parallel-interactive-tutorials) page.

Reviewer: Can you add links to the accompanying notebooks on this page too? Maybe under each section in "Launching Jobs"?

Author: Links added. Thanks for the suggestion. All good to go now.

> For a general introduction to using Jupyter notebooks on Kestrel, please refer to the [official documentation](https://nrel.github.io/HPC/Documentation/Development/Jupyter/). This page covers how to leverage parallel computing with python through Jupyter notebooks on compute nodes. Accompanying notebooks to test if the environment and the job have been configured correctly are under the topic 'Parallel Interactive Tutorials' on the main [Python](../index.md/#parallel-interactive-tutorials) page.

Reviewer: ```suggestion
For a general introduction to using Jupyter notebooks on Kestrel, please refer to the [official documentation](../../../Jupyter/index.md). This page covers how to leverage parallel computing with python through Jupyter notebooks on compute nodes. Accompanying notebooks to test if the environment and the job have been configured correctly are under the topic 'Parallel Interactive Tutorials' on the main [Python](../index.md/#parallel-interactive-tutorials) page.
```
```suggestion
For a general introduction to using Jupyter notebooks on Kestrel, please refer to the [official documentation](https://nrel.github.io/HPC/Documentation/Development/Jupyter/). This page covers how to leverage parallel computing with python through Jupyter notebooks on compute nodes. Accompanying notebooks to test if the environment and the job have been configured correctly are under the topic 'Parallel Interactive Tutorials' on the main [Python](../index.md/#parallel-interactive-tutorials) page.
```

## Setting up your account

Login to Kestrel
```
$ ssh -X <username>@kestrel-gpu.hpc.nrel.gov
```

> Navigate to you project directory

Reviewer: ```suggestion
Navigate to your /projects directory
```

## Install packages

CuPy : [“An open-source array library for GPU-accelerated computing with Python”](https://cupy.dev/)
```
$ nvcc –version
$ conda install -c conda-forge cupy
$ conda install -c conda-forge cupy cudnn cutensor nccl
```

numba-cuda : [“CUDA GPU programming by directly compiling a restricted subset of Python code into CUDA kernels”](https://nvidia.github.io/numba-cuda/user/index.html)
```
$ conda install -c conda-forge numba-cuda
$ conda install -c conda-forge cuda-nvcc cuda-nvrtc "cuda-version>=12.0"
```

mpi4py : [“Python bindings for the Message Passing Interface (MPI) standard to exploit multiple processors”](https://mpi4py.readthedocs.io/en/stable/) 
```
$ conda install -c conda-forge mpi4py openmpi
$ conda install cuda-cudart cuda-version=12
```

ipyparallel : [“Interactive Parallel Computing with IPython”](https://ipyparallel.readthedocs.io/en/latest/)
```
$ conda install ipyparallel
```

Dask : [“A flexible open-source Python library for parallel and distributed computing”](https://www.dask.org/)
```
$ conda install dask
$ conda install dask-jobqueue
$ conda install graphviz
$ conda install ipycytoscape
$ conda install matplotlib
```

## Launching jobs

> A general guide to running jobs on Kestrel can be found in the [official documentation](https://nrel.github.io/HPC/Documentation/Systems/Kestrel/Running/). Below are example procedures suitable for running jobs involving specific python modules, depending on their parallelization capability.

Reviewer: ```suggestion
A general guide to running jobs on Kestrel can be found in the [official documentation](../../../../Systems/Kestrel/Running/index.md). Below are example procedures suitable for running jobs involving specific python modules, depending on their parallelization capability.
```

The text in the red box shows an example of the output parameter `<nodename>` and the text in the yellow box shows an example of the output parameter `<alphabet soup>`, relevant to the following tutorial.

<!-- <img src="metadata/nodeName.png" alt="<nodename>" width="300"/>

<img src="metadata/alphabetSoup.png" alt="<alphabet soup>" width="600"/> -->

![<nodename>](metadata/nodeName.png "nodename"){width=400}
![<alphabet soup>](metadata/alphabetSoup.png "alphabet soup")

<!-- ![<alphabet soup>](metadata/alphabetSoup.png "<alphabet soup>") -->

### GPU compatible modules: E.g. CuPy, numba-cuda etc.

1. Kestrel: Launch an interactive job
    ```
    $ salloc -A <projectname> -t 00:15:00 --partition=debug --gres=gpu:1
    $ module load anaconda3/2024.06.1
    $ conda activate ./myEnv
    $ jupyter-lab --no-browser --ip=$(hostname -s)
    ```

2. Local terminal: Establish a SSH tunnel
    ```
    $ ssh -N -L 8888:<nodename>:8888 <username>@kestrel-gpu.hpc.nrel.gov
    ```

3. Web browser
    ```
    http://127.0.0.1:8888/?token=<alphabet soup>
    ```

    File > New > Notebook > myEnvJupyter

> ### Multi thread capable modules: E.g. Dask

Reviewer: ```suggestion
### Multithread capable modules: E.g. Dask
```

> 1. Kestrel: Launch a multi thread interactive job

Reviewer: ```suggestion
1. Kestrel: Launch a multithread interactive job
```

    ```
    $ salloc -A <projectname> -t 00:15:00 --nodes=1 --ntasks-per-node=104 --partition=debug
    $ module load anaconda3/2024.06.1
    $ conda activate ./myEnv
    $ jupyter-lab --no-browser --ip=$(hostname -s)
    ```

2. Local terminal: Establish a SSH tunnel
    ```
    $ ssh -N -L 8888:<nodename>:8888 <username>@kestrel.hpc.nrel.gov
    ```

3. Web browser
    ```
    http://127.0.0.1:8888/?token=<alphabet soup>
    ```

    File > New > Notebook > myEnvJupyter

> ### Multi node capable job. E.g. mpi4py through ipyparallel 

Reviewer: ```suggestion
### Multinode capable job. E.g. mpi4py through ipyparallel 
```

> 1. Kestrel: Launch a multi node interactive job

Reviewer: ```suggestion
1. Kestrel: Launch a multinode interactive job
```

    ```
    $ salloc -A <projectname> -t 00:15:00 --nodes=2 --ntasks-per-node=1 --partition=short
    $ module load anaconda3/2024.06.1
    $ conda activate ./myEnv
    $ jupyter-lab --no-browser --ip=$(hostname -s)
    ```

2. Local terminal: Establish a SSH tunnel
    ```
    $ ssh -N -L 8888:<nodename>:8888 <username>@kestrel.hpc.nrel.gov
    ```

3. Web browser
    ```
    http://127.0.0.1:8888/?token=<alphabet soup>
    ```

    File > New > Notebook > myEnvJupyter

> ### GPU + multi node jobs. E.g. CuPy + mpi4py through ipyparallel

Reviewer: ```suggestion
### GPU + multinode jobs. E.g. CuPy + mpi4py through ipyparallel
```
